memory checking and profiling changes.

gandrewstone opened the pull request:
This patch adds the option to enable gperftools (heap profiler) to the build.
It also adds a RPC call that dumps the sizes of global structures at runtime.

It also cleans up some known 1 time allocations when Bitcoin shuts down, so the memory check results is easier to understand, fixes a rare socket-left-open issue, and moves some side-effecting code out of an assert() (even though asserts are always "on", let's follow normal assert conventions, in case we ever choose to turn them off in release mode)

gandrewstone committed:
add gperftools memory checking tool, and RPC that dumps sizes
add RPC call that dumps sizes of global data structures
respect fLogIPs
indicate in the docs that this is only available (right now) on Linux

sickpig commented on src/netbase.cpp:
```diff
@@ -330,7 +330,10 @@ static bool Socks5(const std::string& strDest, int port, const ProxyCredentials
+        {
+            CloseSocket(hSocket);
             return error("Proxy username or password too long");
+        }
```
seems unrelated for the scope of this PR

gandrewstone replied:
I'll move this and the assert() fixes into its own PR

sickpig commented on src/requestManager.cpp:
```diff
@@ -131,8 +131,8 @@ void CRequestManager::AskFor(const CInv& obj, CNode* from, int priority)
-      data.AddSource(from);
-      LogPrint("blk", "%s available at %s\n", obj.ToString().c_str(), from->addrName.c_str());
+      if (data.AddSource(from))
+        LogPrint("blk", "%s available at %s\n", obj.ToString().c_str(), from->addrName.c_str());
```
nit: didn't we decided to always use curly brackets?

gandrewstone replied:
IDK that we did.  I thought it was part of Allman style, but I guess its not.  Allman just covers placement of parens.  Personally I don't mind them.  But we should argue this out and update developer-notes.md.  Anyway I will fix here.

sickpig commented on src/stat.h:
```diff
@@ -216,7 +220,7 @@ CStatHistory(const std::string& name, unsigned int operation=STAT_OP_SUM):CStat<
-  ~CStatHistory()
+  virtual ~CStatHistory()
     {
     }
```
fix brackets alignment.

ftrader commented on src/unlimited.cpp:
```diff
@@ -1409,8 +1422,76 @@ UniValue getstat(const UniValue& params, bool fHelp)
+#ifdef DEBUG
+#ifdef DEBUG_LOCKORDER
+extern std::map<std::pair<void*, void*>, LockStack> lockorders;
+//boost::thread_specific_ptr<LockStack> lockstack;
```
dead (commented) code

sickpig commented:
one last point the new RPC call should also add a unit test for it.

greatwolf commented on src/requestManager.h:
```diff
@@ -57,12 +57,13 @@ class CUnknownObj
 class CRequestManager
 {
-  protected:
+    //protected:
```
Why is this here?

gandrewstone replied:
good catch, its was temporary, thanks!

greatwolf commented on src/requestManager.cpp:
```diff
@@ -286,7 +286,7 @@ CNodeRequestData::CNodeRequestData(CNode* n)
-void CUnknownObj::AddSource(CNode* from)
+bool CUnknownObj::AddSource(CNode* from)
 {
   if (std::find_if(availableFrom.begin(), availableFrom.end(), IsCNodeRequestDataThisNode(from)) == availableFrom.end())  // node is not in the request list
```
why not prefer the form `begin(availableFrom), end(availableFrom)`?
this line is also getting too long to the right.
brace alignment needs fixing here too.

gandrewstone replied:
once the std:: is prefixed it becomes even longer.
I don't want to fix the braces as part of this PR.  Looks like the "janitor" is needed here as well as in the dev branch

greatwolf commented on src/unlimited.cpp:
```diff
@@ -637,12 +640,12 @@ void UnlimitedSetup(void)
-	new CStatHistory<uint64_t >("net/send/msg/" +  *i);  // This "leaks" in the sense that it is never freed, but is intended to last the duration of the program.
-      }
+    {
+        mallocedStats.push_front(new CStatHistory<uint64_t >("net/recv/msg/" +  *i));
```
It's worth mentioning you can still shove this into a `unique_ptr` even if it is meant to last the program duration. In this case cleanup happens on shutdown, just either make it `static` or place it in global scope.

gandrewstone replied:
the problem is that I don't really know how many message types there will be (although I guess that this should be discoverable at compile time).  if I did know that I could use an array of unique_ptrs as you suggest, but I'm not sure if it makes much difference.

greatwolf commented on src/unlimited.cpp:
```diff
@@ -660,6 +663,16 @@ void UnlimitedSetup(void)
+        delete obj;
+        mallocedStats.pop_front();
+    }
+}
```
Prefer `= nullptr;`. Also see my comment above on `unique_ptr` which could make `UnlimitedCleanup` unnecessary unless the intent is to log this for statistical purposes?

gandrewstone replied:
"release" is not c++11

greatwolf commented on src/unlimited.h:
```diff
@@ -196,6 +201,7 @@ extern CTweak<uint64_t> blockSigopsPerMb;
 extern CTweak<uint64_t> coinbaseReserve;
 extern CTweak<uint64_t> blockMiningSigopsPerMb;
 
+extern std::list<CStatBase*> mallocedStats;
```
What about just using a `stack` or `vector` for this? `push_back` `pop_back` exist for `vector` so you can still preserve cleanup order.

gandrewstone replied:
does it matter?  I suppose a vector would be smaller.  What other advantages are there?

greatwolf replied:
Structurally it's simpler than a `list`, the addresses are guaranteed to be saved next to each other, could be useful later on if you want to examine the memory. It also removes the need to `#include <list>` which means less work for the compiler.

gandrewstone replied:
I hear you but I don't think that these issues are important enough for me spend the time to switch from list to vector.  If you'd like to switch it, go ahead!  You can check out my branch in your repo, and then issue a PR into my branch.  Or after this gets merged, do it the separately.

greatwolf commented on src/requestManager.cpp:
```diff
@@ -297,15 +297,17 @@ void CUnknownObj::AddSource(CNode* from)
+      }
       availableFrom.push_back(req);
+      return true;
     }
```
I know this part is mostly formatting and updating return type to fit method but this raw loop should be refactored into a `find_if` or using some std algorithm.

gandrewstone replied:
maybe a separate PR.  This PR just changes the function to return true/false to clean up some unnecessarily verbose logging

gandrewstone commented:
note, pulled a few changes into their own PR and squashed down to 1 commit.

greatwolf commented on src/unlimited.cpp:
```diff
@@ -1491,6 +1491,45 @@ extern UniValue getsizes(const UniValue& params, bool fHelp)
+
+    LOCK(cs_vNodes);
+    std::vector<CNode*>::iterator n;
+    ret.push_back(Pair("vNodes",vNodes.size()));
```
needs a space after comma.

greatwolf commented on src/unlimited.cpp:
```diff
@@ -1491,6 +1491,45 @@ extern UniValue getsizes(const UniValue& params, bool fHelp)
 #endif
+
+    LOCK(cs_vNodes);
+    std::vector<CNode*>::iterator n;
```
why scope `n` outside the loop?

greatwolf commented on src/unlimited.cpp:
```diff
@@ -1491,6 +1491,45 @@ extern UniValue getsizes(const UniValue& params, bool fHelp)
+    uint64_t totalThinBlockSize = 0;
+    int disconnected = 0;  // watch # of disconnected nodes to ensure they are being cleaned up
+    for (n = vNodes.begin(); n != vNodes.end(); ++n)
+    {
```
You can reduce the lexical noise somewhat from `(*n)->..` to `n.` by declaring it as a `CNode &`.

greatwolf commented on src/unlimited.cpp:
```diff
@@ -1491,6 +1491,45 @@ extern UniValue getsizes(const UniValue& params, bool fHelp)
+        node.push_back(Pair("vInventoryToSend", (*n)->vInventoryToSend.size()));
+        node.push_back(Pair("setAskFor", (*n)->setAskFor.size()));
+        node.push_back(Pair("mapAskFor", (*n)->mapAskFor.size()));
+        //node.push_back(Pair("", (*n)->.size()));
```
I think you forgot to remove this boilerplate `push_back`.

greatwolf commented:
Is clang-format currently being run as part of the workflow/build process? That will help make many of the formatting nitpicks a thing of the pass.

gandrewstone commented:
Taking this out of WIP.

clang-format is being run in dev and we will backport this functionality to release (@greatwolf would you like to take this on)?  Except for egregious cases, please let clang-format do its thing rather than note every small issue.  Look for substantive issues like "did I forget a structure?"

ftrader commented on src/requestManager.cpp:
```diff
@@ -131,8 +131,10 @@ void CRequestManager::AskFor(const CInv& obj, CNode* from, int priority)
-      LogPrint("blk", "%s available at %s\n", obj.ToString().c_str(), from->addrName.c_str());
+      if (data.AddSource(from))
+      {
+          LogPrint("blk", "%s available at %s\n", obj.ToString().c_str(), from->addrName.c_str());
```
This outputs peer IP address - guard with `logips` check?

gandrewstone replied:
fixed

ftrader commented on src/unlimited.cpp:
```diff
@@ -1409,8 +1422,115 @@ UniValue getstat(const UniValue& params, bool fHelp)
+    ret.push_back(Pair("xpeditedBlk",xpeditedBlk.size()));
+    ret.push_back(Pair("xpeditedBlkUp",xpeditedBlkUp.size()));
+    ret.push_back(Pair("xpeditedTxn",xpeditedTxn.size()));
+    //ret.push_back(Pair("lockstack",lockstack.size()));
```
not needed? perhaps rather document why instead of the commented code.

gandrewstone replied:
fixed

ftrader commented on doc/developer-notes.md:
```diff
@@ -124,6 +124,23 @@ can be very difficult to track down. Compiling with -DDEBUG_LOCKORDER (configure
 
+**Memory Profiling**
+
+Bitcoin Unlimited can be compiled with the libtcmalloc allocation library and
```
The configuration seems to only enable it for Linux at the moment. Maybe mentioned "can be compiled on Linux with the ..." .

It seems as if the library is [available on other platforms](https://github.com/gperftools/gperftools) - I suppose it's been enabled only for Linux initially due to testing resource scarcity.

gandrewstone replied:
yes I haven't tested on other platforms.  This is debug/diagnostic code -- we should only focus on the platforms we will actively run this on.

ftrader replied:
Adding a '... on Linux ...' in the documentation makes it clearer to people running on other platforms that this diagnostic feature isn't currently supported on their platforms.
That's why I raised this point.

ftrader reviewed (commented):
I have not tested this yet, but will try to do so over the next days.

One aspect I'm interested in is how to select / make sure that all the necessary structures are selected in getstructuresizes() .

ftrader commented on src/unlimited.cpp:
```diff
@@ -1409,8 +1422,114 @@ UniValue getstat(const UniValue& params, bool fHelp)
+
+    LOCK(cs_vNodes);
+    std::vector<CNode*>::iterator n;
+    ret.push_back(Pair("vNodes", vNodes.size()));
```
Duplicate output of 'vNodes' - cf. line 1482

ftrader reviewed (changes requested):
I tested this PR on Ubuntu 16.10 x86_64.

It seems hard to determine whether documentation about the perftools is still up to date - the official docs I found were several years old.

Nevertheless, I was able to build, do a heap check run and a profiling run, and request the getstructuresizes data.

I had a look at the Makefile changes to verify that libtcmalloc was linked last (which it needs to be otherwise the heap checking may not work reliably).

So it looks ok to me - can you just fix the duplicate "vNodes" output in getstructuresizes please.

gandrewstone commented:
I fixed the duplicate vNodes and added a note that this is for linux only and squashed the commits in preparation for merging.  I will take ftrader's comment as a signoff as soon as the automated checks validate my squash commit.
